### PR TITLE
platform: add WireGuard support

### PR DESCRIPTION
**Pull request by jbeta**
Initial support for [WireGuard](https://www.wireguard.com/). This code enables NetworkManager to query existing WireGuard links via Generic Netlink and parse them into platform objects.

(Brought over from the mailing list for the NM team's convenience.)

### Previous patchsets
- v3: https://mail.gnome.org/archives/networkmanager-list/2018-June/msg00035.html
- v2: https://mail.gnome.org/archives/networkmanager-list/2018-March/msg00043.html
- v1: https://mail.gnome.org/archives/networkmanager-list/2018-March/msg00009.html

### CODE REVIEW

**Comment by vhumpa** [comment]
NM CI: Please confirm test execution over this PR. Comment .*ok\W+to\W+test.* to confirm and/or .*add\W+to\W+whitelist.* to add the PR author in whitelist for future.

**Comment by zx2c4** [comment]
Looking forward to this being merged, so that we can start on the more interesting part of _configuring_ WireGuard from NM.

**Comment by thom311** [comment]
thanks for the pull request. Unfortunately, this branch conflicts with pr #120, so one of them will have to be rebased. Since this branch is smaller and simpler, let's get the other one first. If you want, one of us can rebase this branch afterwards. Anyway, should be only a matter of a few days...

**Comment by jbeta** [comment]
No worries, I'll rebase after. In the meantime I can keep building from here if this looks OK.

File: src/platform/nm-linux-platform.c [hunk 1]
```diff
@@ -7094,6 +7096,13 @@ constructed (GObject *_object)
 	                                (EVENT_CONDITIONS | ERROR_CONDITIONS | DISCONNECT_CONDITIONS),
 	                                 event_handler, platform);
 
+	priv->genlh = nl_socket_alloc ();
```
Comment (jbeta) [inline src/platform/nm-linux-platform.c]: This is missing a matching `nl_socket_free` on `finalize`. Will fix when rebasing.

File: src/platform/nm-linux-platform.c [hunk 2]
```diff
@@ -181,6 +181,40 @@ G_STATIC_ASSERT (RTA_MAX == (__RTA_MAX - 1));
 
 /*****************************************************************************/
 
+#define WG_CMD_GET_DEVICE 0
```
[the remark this answers is not in the document]
Comment (jbeta) [reply src/platform/nm-linux-platform.c]: Of course, if that's OK with you guys.

In general I was trying to stay close the overall style and code layout of the macsec patches, in this case 85103656e97bedf4cdd80c7eb1832b59f4a29686.

File: src/platform/nm-linux-platform.c [hunk 3]
```diff
@@ -1952,6 +2180,26 @@ _new_from_nl_link (NMPlatform *platform, const NMPCache *cache, struct nlmsghdr
...
 	}
 
+	if (obj->link.type == NM_LINK_TYPE_WIREGUARD) {
+		nm_auto_nmpobj NMPObject *lnk_data_now;
+
+		/* The WireGuard kernel module does not yet send link update
+		 * notifications, so we don't actually update the cache. For
+		 * now, always refetch link data here. */
+		lnk_data_now = nmp_object_new (NMP_OBJECT_TYPE_LNK_WIREGUARD, NULL);
+		if (!_wireguard_get_link_properties (platform, &obj->link, lnk_data_now)) {
```
[the remark this answers is not in the document]
Comment (jbeta) [reply src/platform/nm-linux-platform.c]: That came to mind, but didn't want to mess with `ext_data` after rebasing.

I'll go with whatever you guys decide here.

File: src/platform/nm-linux-platform.c [hunk 4]
```diff
@@ -1690,6 +1725,197 @@ _parse_lnk_vxlan (const char *kind, struct nlattr *info_data)
 
 /*****************************************************************************/
```
[the remark this answers is not in the document]
Comment (jbeta) [reply src/platform/nm-linux-platform.c]: Yeah just following existing style - I'll do that instead.

File: src/platform/nm-platform.h [hunk 5]
```diff
@@ -723,6 +723,13 @@ typedef struct {
 	bool l3miss:1;
```
[the remark this answers is not in the document]
Comment (jbeta) [reply src/platform/nm-platform.h]: Will do.

File: src/platform/nm-linux-platform.c [hunk 6]
```diff
@@ -484,6 +484,18 @@ _support_rta_pref_get (void)
 	return _support_rta_pref >= 0;
 }
 
+/*****************************************************************************
+ * Support Generic Netlink family
+ *****************************************************************************/
+
+static int
+_support_genl_family (struct nl_sock *genlh, const char *name)
```
[the remark this answers is not in the document]
Comment (jbeta) [reply src/platform/nm-linux-platform.c]: That's how the `_support*` family of checks in `nm-linux-platform.c` are supposed to behave, right?

I really don't have any attachment to this function - I'll just drop it if you don't feel it fits with the others.

**Comment by jbeta** [comment]
@lkundrak: thanks for your review!

On the NEWS file entry - I am planning to provide one immediately after some substantial level of support (actual configurability, CLI and such) is sent for merging.

**Comment by zx2c4** [comment]
> On the NEWS file entry - I am planning to provide one immediately after some substantial level of support (actual configurability, CLI and such) is sent for merging.

In other words, this PR is the first in a couple @jbeta has planned for adding full WireGuard support. Looks like the goal now is just to get these fundamentals ironed out first.

File: src/platform/nm-linux-platform.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (thom311) [reply src/platform/nm-linux-platform.c]: when the branch was originally done, `ext_data` didn't even exist, so the question didn't come up.

I think, `NMPObjectLink.ext_data` and `NMPObjectLink.netlink.lnk` share some similarities, in that they are additional data that is attached to the `NMPObjectLink` data (there is also `NMPObjectLink.udev.device`). Optimally, the NMPObject would be only stupid data that we cache, usually obtained via netlink. In practice, that was never the case, because `NMPObjectLink.udev.device`. I think it makes actually sense, to attach all relevant bits to the `NMPObjectLink` objects, including the udev device, the ext-data, and the lnk-data. But while `NMPObjectLink` contains pointers to logic (udev-device, ext-data), the logic itself is not in `NMPObjectLink` itself, and that's the important part.

As it is, all fields in `NMP_OBJECT_TYPE_LNK_WIREGUARD` are plain data, and it thus fits well as it is. I don't think it's worth to convert NMPObjectLnkWireguard into an NMPExtDataWireguard, because the only logic associated with it, is `_wireguard_get_link_properties()`. `_wireguard_get_link_properties()` looks fine to me as it is.

Maybe the only thing that should indeed associated with the `NMPObjectLink` instance (alongside the ext-data), is the `wireguard_family_id`. By caching it in NMLinuxPlatformPrivate, it won't be correctly re-fetched, when the wireguard module is unloaded. By caching it inside NMPObjectLink, the lifetime of the `wireguard_family_id` is correctly limited to the lifetime of the interface (which cannot continue to exist after the module is unloaded).

Long story short, I think it is fine as is, but we should improve caching of `wireguard_family_id` in a future commit.

File: src/platform/nm-linux-platform.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (thom311) [reply src/platform/nm-linux-platform.c]: Not sure that is necessary or simpler :)

File: src/platform/nm-linux-platform.c [hunk 7]
```diff
@@ -3068,6 +3316,10 @@ typedef struct {
 
 		gint is_handling;
 	} delayed_action;
+
+	GHashTable *wifi_data;
```
Comment (thom311) [inline src/platform/nm-linux-platform.c]: unused field

Comment (jbeta) [reply src/platform/nm-linux-platform.c]: That's just a mistake while rebasing - I can drop and update this PR.

File: src/platform/nm-linux-platform.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbeta) [reply src/platform/nm-linux-platform.c]: > Long story short, I think it is fine as is, but we should improve caching of wireguard_family_id in a future commit.

OK.

**Comment by zx2c4** [comment]
@lkundrak - Merge time?
